backend/tournesol/views/unconnected_entities.py

```diff
@@ -27,42 +27,39 @@ class UnconnectedEntitiesView(
     serializer_class = EntityNoExtraFieldSerializer
     permission_classes = [IsAuthenticated]
 
-    _already_visited_node_ = set()
-    _all_connections_ = defaultdict(set)
-
-    def get_related_entities(self, entity_id):
-
-        self._already_visited_node_.add(entity_id)
-
-        related_entities = set()
-        if self._all_connections_.get(entity_id) is None:
-            return set()
+    def get_queryset(self):
+        # Get related entities from source entity
+        source_node = get_object_or_404(Entity, uid=self.kwargs.get("uid"))
+        comparison_pairs = list(
+            Comparison.objects.filter(
+                poll=self.poll_from_url, user=self.request.user
+            ).values_list("entity_1_id", "entity_2_id")
+        )
 
-        for other_node in self._all_connections_.get(entity_id):
-            if other_node not in self._already_visited_node_:
-                related_entities.update(self.get_related_entities(other_node))
+        all_connections = defaultdict(set)
+        already_visited_nodes = set()
 
-        return self._all_connections_.get(entity_id).union(related_entities)
+        for (entity_1_id, entity_2_id) in comparison_pairs:
+            all_connections[entity_1_id].add(entity_2_id)
+            all_connections[entity_2_id].add(entity_1_id)
 
-    def get_queryset(self):
-        # Get related entities from source
-        source_node = Entity.objects.none()
+        def get_related_entities(entity_id):
+            already_visited_nodes.add(entity_id)
+            related_entities = set()
+            if entity_id not in all_connections:
+                return set()
 
-        source_node = get_object_or_404(Entity, uid=self.kwargs.get("uid"))
-        comparisons = list(Comparison.objects.filter(
-            poll=self.poll_from_url, user=self.request.user
-        ))
+            for other_node in all_connections[entity_id]:
+                if other_node not in already_visited_nodes:
+                    related_entities.update(get_related_entities(other_node))
```

Comment:
`entity_id` should be part of related entities. Right now, it is part of it only because comparisons are bi-directional. That works, but I'd prefer the below implementation

```Python
        def get_related_entities(entity_id):
            already_visited_nodes.add(entity_id)
            related_entities = {entity_id}

            for other_node in all_connections.get(entity_id, []):
                if other_node not in already_visited_nodes:
                    related_entities.update(get_related_entities(other_node))

            return related_entities
```

Reply:
I agree. After thinking about it a bit more, I also replaced the recursion with a `while` loop. The recursion limit is quite low in Python (1000 by default in most cases).